host/contracts.go

```diff
@@ -0,0 +1,312 @@
+	txn := formContractReq.Transactions[len(formContractReq.Transactions)-1]
+	hostCollateral := fc.ValidHostOutput.Value.Sub(settings.ContractFee)
+	renterInputs, renterOutputs := len(txn.SiacoinInputs), len(txn.SiacoinOutputs)
+
+	// Fund the formation transaction with the host's collateral.
+	toSign, cleanup, err := sh.wallet.FundTransaction(&txn, hostCollateral, nil)
+	if err != nil {
+		log.Warnln("form contract:", "unable to fund transaction:", err)
+		rpc.WriteResponseErr(stream, errors.New("failed to fund contract formation transaction"))
+		return
+	}
+	defer cleanup()
+
+	hostAdditions := &rhp.RPCFormContractAdditions{
+		Inputs:  txn.SiacoinInputs[renterInputs:],
+		Outputs: txn.SiacoinOutputs[renterOutputs:],
+	}
+
+	// write the host transaction additions.
+	if err := rpc.WriteResponse(stream, hostAdditions); err != nil {
+		log.Warnln("form contract:", "failed to write contract additions:", err)
+		return
+	}
+
+	// read the renter signatures from the stream.
+	var renterSigs rhp.RPCFormContractSignatures
+	if err := rpc.ReadResponse(stream, &renterSigs); err != nil {
+		log.Warnln("form contract:", "failed to read renter signatures:", err)
+		return
+	}
+
+	// build a no-op contract revision.
+	fcr := types.FileContractRevision{
+		Parent: types.FileContractElement{
+			StateElement: types.StateElement{
+				ID: txn.FileContractID(0),
+			},
+			FileContract: fc,
+		},
+		Revision:        fc,
+		RenterSignature: renterSigs.RevisionSignature,
+	}
+	fcr.Revision.RevisionNumber = 1
+
+	// verify the renter's signature
+	if !fc.RenterPublicKey.VerifyHash(vc.ContractSigHash(fcr.Revision), renterSigs.RevisionSignature) {
+		log.Warnln("form contract:", "renter signature is invalid")
+		rpc.WriteResponseErr(stream, errors.New("revision signature is invalid"))
+		return
+	}
+
+	// sign the transaction
+	if err := sh.wallet.SignTransaction(vc, &txn, toSign); err != nil {
+		log.Errorln("form contract:", "failed to sign transaction:", err)
+		rpc.WriteResponseErr(stream, errors.New("failed to sign formation transaction"))
+		return
+	}
+
+	hostSigs := &rhp.RPCFormContractSignatures{
+		ContractSignatures: make([][]types.InputSignature, len(txn.SiacoinInputs)),
+		RevisionSignature:  sh.privkey.SignHash(vc.ContractSigHash(fcr.Revision)),
+	}
+	for i := range txn.SiacoinInputs {
+		hostSigs.ContractSignatures[i] = append(hostSigs.ContractSignatures[i], txn.SiacoinInputs[i].Signatures...)
+	}
+
+	// add the renter's signatures to the transaction
+	for i := range renterSigs.ContractSignatures {
+		txn.SiacoinInputs[i].Signatures = append(txn.SiacoinInputs[i].Signatures, renterSigs.ContractSignatures[i]...)
+	}
+
+	txnset := append(parents, txn)
+
+	/*if err := vc.ValidateTransactionSet(txnset); err != nil {
+		log.Warnln("form contract:", "failed to validate transaction set:", err)
+		rpc.WriteResponse(stream, nil, fmt.Errorf("failed to validate transaction set: %w", err))
+		return
+	}*/
```

Comment:
It should be, but we can't use `vc.ValidateTransactionSet` with the existing stub interfaces in the tests, I'll need to replace the existing stubs with a real validation context and hot wallet.

Reply:
I couldn't decide if I should add a full-on `ghost` style package or not.

Reply:
hmm. I think we should strive to make our tests match real world conditions where possible, so I'm in favor of constructing + validating an actual transaction here (rather than e.g. defining and stubbing a `TransactionValidator` interface).